File: pidev-3a19/src/main/java/Services/DiagnosticAIService.java
package Services;

import model.DiagnosticSuggestion;
import model.Probleme;
import Utils.ImageUploadHelper;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.io.*;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.Base64;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Service d'appel à l'API Hugging Face Router (Qwen) pour suggérer un diagnostic
 * (cause, solution, médicament) à partir de la description du problème.
 * Utilise https://router.huggingface.co (format chat completions compatible OpenAI).
 */
public class DiagnosticAIService {

    private static final String HF_ROUTER_CHAT_URL = "https://router.huggingface.co/v1/chat/completions";
    private static final String PROMPT = "Tu es un expert en problèmes agricoles. À partir de la/les photo(s) et de la description fournies, "
            + "fournis au format JSON strict uniquement, sans texte avant ou après, avec exactement ces clés : "
            + "\"cause\", \"solutionProposee\", \"medicament\". "
            + "cause = cause probable du problème, solutionProposee = solution recommandée, "
            + "medicament = nom du médicament/traitement recommandé et dosage si possible. "
            + "Réponds uniquement avec le JSON, pas de markdown.";

    private final Gson gson = new Gson();
    private final String hfApiKey;
    private final String hfModel;

    public DiagnosticAIService() {
        Properties config = loadConfig();
        this.hfApiKey = getProp(config, "hf.api.key");
        this.hfModel = getProp(config, "hf.model");
    }

    private Properties loadConfig() {
        Properties p = new Properties();
        try (InputStream in = getClass().getResourceAsStream("/config.properties")) {
            if (in != null) p.load(in);
        } catch (IOException ignored) {
        }
        try {
            Path configPath = Path.of(System.getProperty("user.dir")).resolve("config.properties");
            if (Files.exists(configPath)) {
                try (Reader r = Files.newBufferedReader(configPath)) {
                    p.load(r);
                }
            }
        } catch (IOException ignored) {
        }
        return p;
    }

    private static String getProp(Properties p, String key) {
        String v = p.getProperty(key);
        return (v != null && !v.trim().isEmpty()) ? v.trim() : null;
    }

    /**
     * Génère une suggestion de diagnostic à partir du problème (description + photos).
     * Envoie texte et images au format OpenAI (content = tableau image_url + text).
     *
     * @throws RuntimeException si clé API manquante, erreur réseau ou réponse mal formée
     */
    public DiagnosticSuggestion generateFromProbleme(Probleme p) {
        if (hfApiKey == null || hfApiKey.isEmpty()) {
            throw new RuntimeException("Clé Hugging Face manquante. Renseignez hf.api.key dans config.properties.");
        }
        String model = (hfModel != null && !hfModel.isEmpty()) ? hfModel : "Qwen/Qwen3.5-397B-A17B";
        String promptText = "Description du problème : " + (p.getDescription() != null ? p.getDescription() : "") + "\n\n" + PROMPT;

        List<Map<String, Object>> contentList = new ArrayList<>();
        Path uploadsBase = ImageUploadHelper.getBaseUploadDir().getParent();
        if (p.getPhotos() != null && !p.getPhotos().isEmpty()) {
            for (String rel : p.getPhotos().split(";")) {
                String trimmed = rel.trim();
                if (trimmed.isEmpty()) continue;
                Path fullPath = uploadsBase.resolve(trimmed);
                if (!Files.exists(fullPath)) continue;
                try {
                    byte[] bytes = Files.readAllBytes(fullPath);
                    String base64 = Base64.getEncoder().encodeToString(bytes);
                    String mime = "image/jpeg";
                    String lower = trimmed.toLowerCase();
                    if (lower.endsWith(".png")) mime = "image/png";
                    else if (lower.endsWith(".gif")) mime = "image/gif";
                    else if (lower.endsWith(".webp")) mime = "image/webp";
                    String dataUrl = "data:" + mime + ";base64," + base64;
                    contentList.add(Map.of(
                            "type", "image_url",
                            "image_url", Map.of("url", dataUrl)
                    ));
                } catch (IOException e) {
                    throw new RuntimeException("Impossible de charger l'image : " + trimmed, e);
                }
            }
        }
        contentList.add(Map.of("type", "text", "text", promptText));

        Map<String, Object> message = Map.of("role", "user", "content", contentList);
        Map<String, Object> body = new HashMap<>();
        body.put("model", model);
        body.put("messages", List.of(message));
        body.put("max_tokens", 4096);
        body.put("temperature", 0.6);
        body.put("top_p", 0.95);
        body.put("top_k", 20);

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(HF_ROUTER_CHAT_URL))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + hfApiKey)
                .POST(HttpRequest.BodyPublishers.ofString(gson.toJson(body)))
                .build();

        HttpResponse<String> response;
        try {
            response = client.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (IOException e) {
            throw new RuntimeException("Erreur réseau : " + e.getMessage());
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("Appel interrompu");
        }

        if (response.statusCode() != 200) {
            String bodyStr = response.body();
            if (response.statusCode() == 401) {
                throw new RuntimeException("Clé Hugging Face invalide. Vérifiez hf.api.key sur https://huggingface.co/settings/tokens");
            }
            throw new RuntimeException("Hugging Face erreur " + response.statusCode() + " : " + (bodyStr != null && bodyStr.length() > 250 ? bodyStr.substring(0, 250) + "..." : bodyStr));
        }

        String rawBody = response.body();
        String responseText = extractHfRouterResponseText(rawBody);
        if (responseText == null || responseText.isEmpty()) {
            String preview = rawBody != null && rawBody.length() > 400 ? rawBody.substring(0, 400) + "..." : rawBody;
            throw new RuntimeException("Réponse Hugging Face vide ou mal formée. Réponse reçue : " + (preview != null ? preview : "(vide)"));
        }
        return parseSuggestionFromText(responseText);
    }

    /**
     * Extrait le texte de la réponse HF Router.
     * Gère : choices[0].message.content (string ou tableau de parties type/text).
     */
    private String extractHfRouterResponseText(String jsonBody) {
        if (jsonBody == null || jsonBody.isBlank()) return null;
        try {
            JsonObject root = gson.fromJson(jsonBody, JsonObject.class);
            if (root == null || !root.has("choices")) return null;
            var choices = root.getAsJsonArray("choices");
            if (choices == null || choices.size() == 0) return null;
            JsonObject choice = choices.get(0).getAsJsonObject();
            // Format OpenAI : choice.message.content
            if (choice.has("message")) {
                JsonObject message = choice.getAsJsonObject("message");
                JsonElement content = message.get("content");
                if (content == null) return null;
                if (content.isJsonPrimitive() && content.getAsJsonPrimitive().isString()) {
                    return content.getAsString();
                }
                if (content.isJsonArray()) {
                    var parts = content.getAsJsonArray();
                    StringBuilder sb = new StringBuilder();
                    for (int i = 0; i < parts.size(); i++) {
                        JsonObject part = parts.get(i).getAsJsonObject();
                        if (part.has("type") && "text".equals(part.get("type").getAsString()) && part.has("text")) {
                            sb.append(part.get("text").getAsString());
                        }
                    }
                    return sb.length() > 0 ? sb.toString() : null;
                }
            }
            // Certaines APIs : choice.text
            if (choice.has("text")) {
                return choice.get("text").getAsString();
            }
            return null;
        } catch (Exception e) {
            return null;
        }
    }

    private DiagnosticSuggestion parseSuggestionFromText(String text) {
        String json = text.trim();
        Matcher m = Pattern.compile("(?s)```(?:json)?\\s*(.*?)\\s*```").matcher(json);
        if (m.find()) {
            json = m.group(1).trim();
        }
        try {
            JsonObject obj = gson.fromJson(json, JsonObject.class);
            if (obj == null) throw new RuntimeException("JSON invalide");
            String cause = obj.has("cause") ? obj.get("cause").getAsString() : "";
            String solutionProposee = obj.has("solutionProposee") ? obj.get("solutionProposee").getAsString() : "";
            String medicament = obj.has("medicament") ? obj.get("medicament").getAsString() : "";
            return new DiagnosticSuggestion(cause, solutionProposee, medicament);
        } catch (Exception e) {
            throw new RuntimeException("Impossible de parser la réponse IA : " + e.getMessage());
        }
    }
}

File: farmtech-web/src/Service/DiagnosticAIService.php
<?php

namespace App\Service;

use App\Entity\Diagnostique;
use App\Entity\Probleme;
use App\Repository\DiagnostiqueRepository;
use Symfony\Contracts\HttpClient\HttpClientInterface;

/**
 * Suggestion diagnostic via OpenRouter.
 */
final class DiagnosticAIService
{
    private const OPENROUTER_DEFAULT_CHAT_URL = 'https://openrouter.ai/api/v1/chat/completions';

    private const BASE_PROMPT = 'Tu es un expert en problèmes agricoles. À partir de la/les photo(s), de la description et, '
        .'si fourni, de l’historique des révisions et feedbacks fermier, '
        .'propose une suite cohérente (évite de contredire sans raison un diagnostic déjà approuvé sauf si les faits imposent une correction). '
        .'Fournis au format JSON strict uniquement, sans texte avant ou après, avec exactement ces clés : '
        .'"cause", "solutionProposee", "medicament". '
        .'cause = cause probable du problème, solutionProposee = solution recommandée, '
        .'medicament = nom du médicament/traitement recommandé et dosage si possible. '
        .'Réponds uniquement avec le JSON, pas de markdown.';

    private const HISTO_MAX_TOTAL_CHARS = 12000;

    private const HISTO_FIELD_CLIP = 900;

    private const MAX_RETRY_COUNT = 2;

    private const SIMILARITY_THRESHOLD = 0.84;

    public function __construct(
        private readonly HttpClientInterface $httpClient,
        private readonly string $openRouterApiKey,
        private readonly string $openRouterModel,
        private readonly string $openRouterChatUrl,
        private readonly string $projectDir,
        private readonly DiagnostiqueRepository $diagnostiqueRepository,
    ) {
    }

    /**
     * @return array{cause: string, solutionProposee: string, medicament: string}
     */
    public function generateFromProbleme(Probleme $p): array
    {
        if ('' === trim($this->openRouterApiKey)) {
            throw new \RuntimeException('Clé OpenRouter manquante. Renseignez OPENROUTER_API_KEY dans .env.');
        }

        $model = '' !== trim($this->openRouterModel) ? $this->openRouterModel : 'google/gemma-4-26b-a4b-it';
        $desc = $p->getDescription() ?? '';
        $revisions = $this->diagnostiqueRepository->findByProblemeOrderedByRevision($p);
        $revisionCount = \count($revisions);
        $latest = $revisionCount > 0 ? $revisions[$revisionCount - 1] : null;
        $historique = $this->buildHistoriqueContext($revisions);
        $retryHint = null;

        for ($attempt = 0; $attempt <= self::MAX_RETRY_COUNT; ++$attempt) {
            $promptText = $this->buildPromptText($desc, $historique, $revisionCount, $retryHint);
            $raw = $this->requestCompletion($model, $p, $promptText);
            $suggestion = $this->parseSuggestionFromText($raw);
            if (!$latest instanceof Diagnostique || !$this->isTooSimilarToLatest($suggestion, $latest)) {
                return $suggestion;
            }
            $retryHint = 'La réponse précédente ressemble trop à la dernière révision. '
                .'Reformule de manière nettement différente la cause et la solution, en gardant la cohérence technique.';
        }

        throw new \RuntimeException('Impossible de générer une révision suffisamment différente. Réessayez.');
    }

    /**
     * Révisions existantes + feedbacks fermier pour contextualiser la suggestion (tronqué pour limiter les tokens).
     */
    private function buildHistoriqueContext(array $revs): string
    {
        if ($revs === []) {
            return '';
        }
        $blocks = ['--- Historique des révisions et retours fermier ---'];
        $total = 0;
        foreach ($revs as $d) {
            $chunk = $this->formatRevisionBlock($d);
            if ('' === $chunk) {
                continue;
            }
            if ($total + \strlen($chunk) > self::HISTO_MAX_TOTAL_CHARS) {
                $blocks[] = '… (historique tronqué pour limite de taille)';
                break;
            }
            $blocks[] = $chunk;
            $total += \strlen($chunk);
        }

        return implode("\n", $blocks);
    }

    private function buildPromptText(string $description, string $historique, int $revisionCount, ?string $retryHint): string
    {
        $promptText = 'Description du problème : '.$description."\n\n";
        if ('' !== $historique) {
            $promptText .= $historique."\n\n";
        }
        $promptText .= self::BASE_PROMPT."\n";
        if ($revisionCount === 0) {
            $promptText .= 'Ceci est la 1re réponse: ne pas présenter d’excuses et aller droit au diagnostic.';
        } else {
            $promptText .= 'Ceci est une révision (2e ou +): commencer la cause ou la solution par une phrase courte '
                .'d’empathie en français reconnaissant que la solution précédente n’a pas fonctionné, puis proposer '
                .'un plan ajusté et concret.';
        }
        if (null !== $retryHint && '' !== trim($retryHint)) {
            $promptText .= "\n".$retryHint;
        }

        return $promptText;
    }

    private function requestCompletion(string $model, Probleme $probleme, string $promptText): string
    {
        $chatUrl = '' !== trim($this->openRouterChatUrl) ? $this->openRouterChatUrl : self::OPENROUTER_DEFAULT_CHAT_URL;
        $contentList = $this->buildContentList($probleme, $promptText);
        $body = [
            'model' => $model,
            'messages' => [['role' => 'user', 'content' => $contentList]],
            'max_tokens' => 4096,
            'temperature' => 0.7,
            'top_p' => 0.95,
            'top_k' => 20,
        ];

        $response = $this->httpClient->request('POST', $chatUrl, [
            'headers' => [
                'Content-Type' => 'application/json',
                'Authorization' => 'Bearer '.$this->openRouterApiKey,
                'HTTP-Referer' => 'http://localhost:8000',
                'X-Title' => 'farmtech-web',
            ],
            'json' => $body,
            'timeout' => 120,
        ]);

        $status = $response->getStatusCode();
        $rawBody = $response->getContent(false);
        if (401 === $status) {
            throw new \RuntimeException('Clé OpenRouter invalide. Vérifiez OPENROUTER_API_KEY sur https://openrouter.ai/keys');
        }
        if (200 !== $status) {
            $preview = strlen($rawBody) > 250 ? substr($rawBody, 0, 250).'...' : $rawBody;
            throw new \RuntimeException('OpenRouter erreur '.$status.' : '.$preview);
        }
        $data = json_decode($rawBody, true);
        if (\is_array($data)) {
            $text = $this->extractHfRouterResponseText($data);
            if (null !== $text && '' !== trim($text)) {
                return $text;
            }
        }
        $fallback = $this->extractJsonObjectFromRawBody($rawBody);
        if (null !== $fallback) {
            return $fallback;
        }
        $preview = strlen($rawBody) > 250 ? substr($rawBody, 0, 250).'...' : $rawBody;
        throw new \RuntimeException('Réponse Hugging Face vide ou mal formée. Extrait: '.$preview);
    }

    /**
     * @return array<int, array{type:string, text?:string, image_url?:array{url:string}}>
     */
    private function buildContentList(Probleme $probleme, string $promptText): array
    {
        $contentList = [];
        $photos = $probleme->getPhotos();
        if (null !== $photos && '' !== trim($photos)) {
            foreach (explode(';', $photos) as $rel) {
                $trimmed = trim($rel);
                if ('' === $trimmed) {
                    continue;
                }
                $full = $this->projectDir.'/public/'.str_replace('\\', '/', $trimmed);
                if (str_contains($full, '..') || !is_file($full)) {
                    continue;
                }
                $bytes = @file_get_contents($full);
                if (false === $bytes) {
                    continue;
                }
                $mime = 'image/jpeg';
                $lower = strtolower($trimmed);
                if (str_ends_with($lower, '.png')) {
                    $mime = 'image/png';
                } elseif (str_ends_with($lower, '.gif')) {
                    $mime = 'image/gif';
                } elseif (str_ends_with($lower, '.webp')) {
                    $mime = 'image/webp';
                }
                $dataUrl = 'data:'.$mime.';base64,'.base64_encode($bytes);
                $contentList[] = [
                    'type' => 'image_url',
                    'image_url' => ['url' => $dataUrl],
                ];
            }
        }
        $contentList[] = ['type' => 'text', 'text' => $promptText];

        return $contentList;
    }

    /**
     * @param array{cause:string, solutionProposee:string, medicament:string} $candidate
     */
    private function isTooSimilarToLatest(array $candidate, Diagnostique $latest): bool
    {
        $latestCause = $this->normalizeText((string) ($latest->getCause() ?? ''));
        $latestSolution = $this->normalizeText((string) ($latest->getSolutionProposee() ?? ''));
        $newCause = $this->normalizeText($candidate['cause'] ?? '');
        $newSolution = $this->normalizeText($candidate['solutionProposee'] ?? '');
        if ('' === $latestCause || '' === $latestSolution || '' === $newCause || '' === $newSolution) {
            return false;
        }

        return $this->similarityRatio($latestCause, $newCause) >= self::SIMILARITY_THRESHOLD
            && $this->similarityRatio($latestSolution, $newSolution) >= self::SIMILARITY_THRESHOLD;
    }

    private function normalizeText(string $value): string
    {
        $value = strtolower(trim($value));
        if ('' === $value) {
            return '';
        }

        return (string) preg_replace('/\s+/', ' ', $value);
    }

    private function similarityRatio(string $left, string $right): float
    {
        similar_text($left, $right, $percent);

        return $percent / 100.0;
    }

    private function formatRevisionBlock(Diagnostique $d): string
    {
        $lines = [];
        $lines[] = sprintf(
            'Révision n°%d — %s — approuvé pour le fermier : %s',
            $d->getNumRevision(),
            $d->getDateDiagnostique()?->format('Y-m-d H:i') ?? '?',
            $d->isApprouve() ? 'oui' : 'non',
        );
        $lines[] = 'Résultat : '.$this->clipText((string) ($d->getResultat() ?? ''));
        $lines[] = 'Cause : '.$this->clipText((string) ($d->getCause() ?? ''));
        $lines[] = 'Solution proposée : '.$this->clipText((string) ($d->getSolutionProposee() ?? ''));
        $med = $d->getMedicament();
        if (null !== $med && '' !== trim($med)) {
            $lines[] = 'Médicament / traitement : '.$this->clipText($med);
        }
        $fb = $d->getFeedbackFermier();
        if (null !== $fb && '' !== $fb) {
            $label = 'RESOLU' === $fb ? 'Résolu' : ('NON_RESOLU' === $fb ? 'Non résolu' : $fb);
            $lines[] = 'Feedback fermier : '.$label;
            if (null !== $d->getDateFeedback()) {
                $lines[] = 'Date feedback : '.$d->getDateFeedback()->format('Y-m-d H:i');
            }
            $com = $d->getFeedbackCommentaire();
            if (null !== $com && '' !== trim($com)) {
                $lines[] = 'Commentaire fermier : '.$this->clipText($com);
            }
        }
        $lines[] = '---';

        return implode("\n", $lines);
    }

    private function clipText(string $text): string
    {
        $text = trim($text);
        if ('' === $text) {
            return '—';
        }
        if (\strlen($text) <= self::HISTO_FIELD_CLIP) {
            return $text;
        }

        return substr($text, 0, self::HISTO_FIELD_CLIP - 1).'…';
    }

    /**
     * @param array<string, mixed> $root
     */
    private function extractHfRouterResponseText(array $root): ?string
    {
        if (!isset($root['choices'][0])) {
            return null;
        }
        $choice = $root['choices'][0];
        if (!\is_array($choice)) {
            return null;
        }
        if (isset($choice['message']['content'])) {
            $content = $choice['message']['content'];
            if (\is_string($content)) {
                return $content;
            }
            if (\is_array($content)) {
                $sb = '';
                foreach ($content as $part) {
                    if (!\is_array($part)) {
                        continue;
                    }
                    if (isset($part['text']) && \is_string($part['text'])) {
                        $sb .= $part['text'];
                        continue;
                    }
                    if (isset($part['content']) && \is_string($part['content'])) {
                        $sb .= $part['content'];
                    }
                }

                return '' !== $sb ? $sb : null;
            }
        }
        if (isset($choice['message']['function_call']['arguments']) && \is_string($choice['message']['function_call']['arguments'])) {
            return $choice['message']['function_call']['arguments'];
        }
        if (isset($choice['message']['tool_calls']) && \is_array($choice['message']['tool_calls'])) {
            $sb = '';
            foreach ($choice['message']['tool_calls'] as $toolCall) {
                if (!\is_array($toolCall)) {
                    continue;
                }
                if (isset($toolCall['function']['arguments']) && \is_string($toolCall['function']['arguments'])) {
                    $sb .= $toolCall['function']['arguments']."\n";
                }
            }
            if ('' !== trim($sb)) {
                return trim($sb);
            }
        }
        if (isset($choice['message']['reasoning_content']) && \is_string($choice['message']['reasoning_content'])) {
            return $choice['message']['reasoning_content'];
        }
        if (isset($choice['text']) && \is_string($choice['text'])) {
            return $choice['text'];
        }
        if (isset($root['generated_text']) && \is_string($root['generated_text'])) {
            return $root['generated_text'];
        }

        return null;
    }

    private function extractJsonObjectFromRawBody(string $rawBody): ?string
    {
        $trimmed = trim($rawBody);
        if ('' === $trimmed) {
            return null;
        }
        if (str_starts_with($trimmed, '{') && str_contains($trimmed, '"cause"') && str_contains($trimmed, '"solutionProposee"')) {
            return $trimmed;
        }
        if (preg_match('/\{(?:[^{}]|(?R))*"cause"(?:[^{}]|(?R))*"solutionProposee"(?:[^{}]|(?R))*\}/s', $trimmed, $m)) {
            return $m[0];
        }

        return null;
    }

    /**
     * @return array{cause: string, solutionProposee: string, medicament: string}
     */
    private function parseSuggestionFromText(string $text): array
    {
        $payload = trim($text);
        if (preg_match('/(?s)```(?:json)?\s*(.*?)\s*```/', $payload, $m)) {
            $payload = trim($m[1]);
        }
        $obj = $this->decodeSuggestionObject($payload);
        if (!\is_array($obj)) {
            $fallback = $this->parseSuggestionFromLabeledText($payload);
            if (null !== $fallback) {
                return $fallback;
            }
            throw new \RuntimeException('Impossible de parser la réponse IA.');
        }
        $cause = $this->firstStringByKeys($obj, ['cause', 'causes', 'rootCause']);
        $solution = $this->firstStringByKeys($obj, ['solutionProposee', 'solution_proposee', 'solution', 'planAction']);
        $medicament = $this->firstStringByKeys($obj, ['medicament', 'traitement', 'treatment']);

        return [
            'cause' => $cause ?? '',
            'solutionProposee' => $solution ?? '',
            'medicament' => $medicament ?? '',
        ];
    }

    /**
     * @return array<string,mixed>|null
     */
    private function decodeSuggestionObject(string $payload): ?array
    {
        $try = trim($payload);
        $obj = json_decode($try, true);
        if (\is_array($obj)) {
            return $obj;
        }
        if (preg_match('/\{(?:[^{}]|(?R))*\}/s', $try, $m)) {
            $try = $m[0];
            $obj = json_decode($try, true);
            if (\is_array($obj)) {
                return $obj;
            }
        }
        $normalized = str_replace(["\u{201C}", "\u{201D}", "\u{2018}", "\u{2019}"], ['"', '"', "'", "'"], $try);
        $normalized = (string) preg_replace('/,\s*([}\]])/', '$1', $normalized);
        $obj = json_decode($normalized, true);
        if (\is_array($obj)) {
            return $obj;
        }

        return null;
    }

    private function firstStringByKeys(array $obj, array $keys): ?string
    {
        foreach ($keys as $key) {
            if (isset($obj[$key]) && \is_string($obj[$key]) && '' !== trim($obj[$key])) {
                return trim($obj[$key]);
            }
        }

        return null;
    }

    /**
     * @return array{cause:string, solutionProposee:string, medicament:string}|null
     */
    private function parseSuggestionFromLabeledText(string $payload): ?array
    {
        $cause = '';
        $solution = '';
        $medicament = '';
        $lines = preg_split('/\R+/', trim($payload)) ?: [];
        foreach ($lines as $line) {
            $line = trim((string) $line);
            if ('' === $line || !str_contains($line, ':')) {
                continue;
            }
            [$label, $value] = array_pad(explode(':', $line, 2), 2, '');
            $label = strtolower(trim($label));
            $value = trim($value);
            if ('' === $value) {
                continue;
            }
            if (str_contains($label, 'cause') && '' === $cause) {
                $cause = $value;
                continue;
            }
            if ((str_contains($label, 'solution') || str_contains($label, 'plan')) && '' === $solution) {
                $solution = $value;
                continue;
            }
            if ((str_contains($label, 'médicament') || str_contains($label, 'medicament') || str_contains($label, 'traitement')) && '' === $medicament) {
                $medicament = $value;
            }
        }
        if ('' === $cause && '' === $solution && '' === $medicament) {
            return null;
        }

        return ['cause' => $cause, 'solutionProposee' => $solution, 'medicament' => $medicament];
    }
}
